zkevm-circuits/src/exp_circuit.rs

```diff
@@ -0,0 +1,569 @@
+                        meta.query_advice(mul_gadget_col, Rotation(1)),
+                    );
+                }
+            });
+
+            cb.gate(meta.query_selector(q_step))
+        });
+
+        Self {
+            q_step,
+            is_pad,
+            exp_table,
+            mul_gadget,
+            parity_check,
+        }
+    }
+
+    /// Assign witness to the exponentiation circuit.
+    pub fn assign_block(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        block: &Block<F>,
+    ) -> Result<(), Error> {
+        const OFFSET_INCREMENT: usize = 7usize;
+        const ROWS_PER_STEP: usize = 4usize;
+
+        let mul_chip = MulAddChip::construct(self.mul_gadget.clone());
+        let parity_check_chip = MulAddChip::construct(self.parity_check.clone());
+
+        layouter.assign_region(
+            || "exponentiation circuit",
+            |mut region| {
+                // assign everything except the exp table.
+                let mut offset = 0;
+                for exp_event in block.exp_events.iter() {
+                    let mut exponent = exp_event.exponent;
+                    for step in exp_event.steps.iter().rev() {
+                        let two = U256::from(2);
+                        let (exponent_div2, remainder) = exponent.div_mod(two);
+
+                        self.q_step.enable(&mut region, offset)?;
+                        region.assign_advice(
+                            || format!("is_pad: {}", offset),
+                            self.is_pad,
+                            offset,
+                            || Value::known(F::zero()),
+                        )?;
+                        mul_chip.assign(
+                            &mut region,
+                            offset,
+                            [step.a, step.b, U256::zero(), step.d],
+                        )?;
+                        parity_check_chip.assign(
+                            &mut region,
+                            offset,
+                            [two, exponent_div2, remainder, exponent],
+                        )?;
+
+                        // update reducing exponent
+                        if remainder.is_zero() {
+                            // exponent is even
+                            exponent = exponent_div2;
+                        } else {
+                            // exponent is odd
+                            exponent = exponent - 1;
+                        }
+
+                        // mul_chip has 7 rows, exp_table has 4 rows. So we increment the offset by
+                        // the maximum number of rows taken up by any gadget within the
+                        // exponentiation circuit.
+                        offset += OFFSET_INCREMENT;
+                    }
+                    // Additionally, 1 row is reserved for a padding row at the end of every
+                    // exponentiation trace.
+                    offset += 1;
```

Comment:
It doesn't matter if the `q_step == 1` comes at different offsets. Because the gate is enabled only where `q_step == 1`. And wherever `q_step == 1`, then each step takes 7 rows. So for the scenario you mention, there shouldn't be an issue.